src/main/java/org/apache/brooklyn/cloudfoundry/location/CloudFoundryClientRegistryImpl.java

```diff
@@ -41,23 +39,30 @@ protected CloudFoundryClientRegistryImpl() {
     public CloudFoundryOperations getCloudFoundryClient(ConfigBag conf, boolean allowReuse) {
         String username = checkNotNull(conf.get(PaasLocationConfig.ACCESS_IDENTITY), "identity must not be null");
         String password = checkNotNull(conf.get(PaasLocationConfig.ACCESS_CREDENTIAL), "credential must not be null");
-        URL apiHost = getTargetURL(checkNotNull(conf.get(PaasLocationConfig.CLOUD_ENDPOINT), "endpoint must not be null"));
+        String apiHost = checkNotNull(conf.get(PaasLocationConfig.CLOUD_ENDPOINT), "endpoint must not be null");
         String organization = checkNotNull(conf.get(CloudFoundryPaasLocationConfig.CF_ORG), "organization must not be null");
         String space = checkNotNull(conf.get(CloudFoundryPaasLocationConfig.CF_SPACE), "space must not be null");
 
-        CloudCredentials credentials = new CloudCredentials(username, password);
-        CloudFoundryClient client = new CloudFoundryClient(credentials, apiHost, organization, space, true);
-        client.login();
+        DefaultConnectionContext connectionContext = DefaultConnectionContext.builder()
+                .apiHost(apiHost)
+                .build();
+        //TODO: using singleton for the class instance
+        PasswordGrantTokenProvider tokenProvider = PasswordGrantTokenProvider.builder()
```

Comment:
The `tokenProvider`  and the `connectionContext` should be created following the singleton pattern, they should be shared for each client, according to the `cf-java-client` [documentation](https://github.com/cloudfoundry/cf-java-client/tree/v2.0.0.M10#cloudfoundryclient-dopplerclient-uaaclient-builders)
@andreaturli, what do you think?